git-submodules-status

```diff
@@ -0,0 +1,106 @@
+#       version, the current version and a small visual cue to easily identify
+#       mismatches between the two.
+#
+#   Note that this script assumes that a project's submodules are present in a
+#   directory called "submodules" in the project's top-level git directory.
+#
+#   Usage:
+#   ------
+#   (first place this script in one of the directories in $PATH)
+#       $> cd path/to/repo
+#       $> git submodules-status
+#
+################################################################################
+
+appendSpaces()
+{
+    INITIAL_STR=$1
+    TARGET_LENGTH=$2
+
+    SPACES_TO_APPEND=$((TARGET_LENGTH - ${#INITIAL_STR}))
+
+    echo -n $INITIAL_STR
+    for ((i = 0; i < $SPACES_TO_APPEND; i++))
+    do
+        echo -n " "
+    done
+    echo ""
+}
+
+# Confirm that we are in a git repository
+GIT_TOP=$(git rev-parse --show-toplevel) || exit $?
+
+# Check if the project has any submodules in the first place
+if ! [ -a $GIT_TOP/.gitmodules ]; then
+    PROJECT_NAME=$(basename $GIT_TOP)
+    echo "Project '$PROJECT_NAME' has no submodules. Nothing to do here!"
+    exit 0
+fi
+
+# Confirm that all the submodules have been initialised
+UNINITIALISED_SUBMODS=(`git submodule | grep -v "^ " | awk -F"/" '{print $NF}'`)
+if [ ${#UNINITIALISED_SUBMODS[@]} -gt 0 ]; then
+    echo "The following submodules haven't been initialized : ${UNINITIALISED_SUBMODS[@]}"
+    echo "Try running this first:"
+    echo "    git submodule update --init"
+    exit 2
+fi
+
+# Read the existing submodules information into an array. After this, the array
+# should contain an even number of entries, with each pair of entries containing
+# the submodule name and the current submodule version respectively.
+ARR=(`git submodule | sed 's|^.*submodules/\(.*\) (\(.*\))$|\1 \2|g'`)
+
+# Fetch each submodule from the remote repo so that we can get its latest tag
+git submodule foreach git remote update -p
+echo ""
+
+# Get the length of the longest submodule name
+MAX_LEN=9 # Initialize to length of the header string "Submodule"
+for ((i = 0; i < ${#ARR[@]}; i+=2))
+do
+    CUR_LEN=${#ARR[$i]}
+
+    if [ $CUR_LEN -gt $MAX_LEN ]
+    then
+        MAX_LEN=$CUR_LEN
+    fi
+done
+
+FIRST_COLUMN_WIDTH=$((MAX_LEN+4))
+
+# For now, the width of the second column is hard-coded to 19
+# i.e. length of the string "Current version" + 4
+SECOND_COLUMN_WIDTH=19
+
+# Also, the width of the third column is hard-coded to 12
+# i.e. length of the string "Latest version" - 2
+# so that the '✗' which indicates a difference between the two versions can
+# appear as the last character of the "Latest version" column
+THIRD_COLUMN_WIDTH=12
+
+# Write the header row and its underlines
+echo "$(appendSpaces "Submodule" $FIRST_COLUMN_WIDTH) Current version     Latest version"
+echo "---------     ---------------     --------------"
+
+# Write each submodule's data
+for ((i = 0; i < ${#ARR[@]}; i+=2))
+do
+    cd $GIT_TOP/submodules/${ARR[$i]}
```

Comment:
This line is another reason for the `submodules` directory restriction. But even this can be dealt with I think.